### DISCUSSION TITLE
py --list --help does not print help message

### DISCUSSION THREAD

**Discussion by gaborbernat** [brettcannon/python-launcher#77] [Ideas]
IMHO the help message should be respected irrespective of where it's specified on the command line.

**Comment by brettcannon** [comment]
So what would you expect `py --help --list` do? And how are you proposing to differentiate flags to the interpreter compared to the launcher? And what Python version are you proposing to have represent the output of `--help` since one wasn't appropriately selected?

FYI the Windows launcher errors out in one case about not recognizing the flag:
```PS
$ py --list --help
unknown option --list
usage: C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.496.0_x64__qbz5n2kfra8p0\python3.9.exe [option] ... [-c cmd | -m mod | file | -] [arg] ...
Try `python -h' for more information.
```

And says it can't find the appropriate version in the other case:
```PS
$ py -0 --help
Python 0 not found!
Installed Pythons found by C:\Users\brcan\AppData\Local\Programs\Python\Launcher\py.exe Launcher for Windows
 -3.9-64 *
 -3.8-64

Requested Python version (0) not installed, use -0 for available pythons
```

**Reply by gaborbernat** [reply]
I'd expect ``py --list --help`` to do the same as ``py --help``. This is what argparse would do. The way I see it you have two options:

- if I pass in any arguments that are consumed by py then all arguments are considered to be for py
- if I pass in any arguments that are consumed by py then as many arguments as they valid are considered to be for py

``--list`` doesn't show up for python, only for py so it's safe to assume --help should be respected at the same time.

> And what Python version are you proposing to have represent the output of `--help` since one wasn't appropriately selected?

In this case show the help part only for ``py``, and not the python part. Not?

**Reply by brettcannon** [reply]
> This is what argparse would do.

But this isn't using argparse, so not so easy to make happen. 😉 

> if I pass in any arguments that are consumed by py then all arguments are considered to be for py

This doesn't work because the Launcher is specifically designed to have minimal arguments and hence why the current design only checks the first argument and ignores the rest to pass on to Python. Basically `py -3 -c pass` is legitimate and so I can't assume `-c` is for the Launcher instead of Python itself.

> if I pass in any arguments that are consumed by py then as many arguments as they valid are considered to be for py

Another option would simply refuse to guess for `--list` if more than one argument is provided as it's the only flag for the Launcher other than `--help` which doesn't expect to actually pass anything to Python. Basically the equivalent of:

```Python
if len(sys.argv) > 2 and sys.argv[1] in {'--list', '--help'}:
    raise SystemExit(f"The {sys.argv[1]} flag is expected to be specified by itself; see `py --help` for details")
```


> In this case show the help part only for py, and not the python part. No?

Don't know. There isn't a concept of _just_ showing the Launcher's help right now.

**Comment by brettcannon** [comment]
I'm going to go with the error scenario. Opened https://github.com/brettcannon/python-launcher/issues/80 to track this.
